File: reference-implementations/java/src/main/java/org/tomlschema/DiscoveredSchema.java
package org.tomlschema;

import org.tomlj.TomlParseResult;

import java.util.List;

/**
 * The outcome of discovering a schema from a TOML document via its reserved
 * {@code [toml-schema].location}, as described in SPEC.md's
 * "TOML Reference of a TOML Schema" section.
 *
 * @param schema the discovered and loaded schema
 * @param document the parsed TOML document that referenced the schema
 * @param warnings non-fatal version-compatibility warnings produced while comparing
 *                  the document's expected TOML Schema version against the resolved
 *                  schema's declared version; empty when the document omits
 *                  {@code [toml-schema].version} or the versions match exactly
 */
public record DiscoveredSchema(TomlSchema schema, TomlParseResult document, List<ValidationWarning> warnings) {
    public DiscoveredSchema {
        warnings = List.copyOf(warnings);
    }

    /**
     * Validates the discovered document against the discovered schema, including any
     * version-compatibility warnings produced during discovery.
     *
     * @return the combined validation result
     */
    public ValidationResult validate() {
        ValidationResult result = schema.validate(document);
        if (warnings.isEmpty()) {
            return result;
        }
        List<ValidationWarning> combined = new java.util.ArrayList<>(warnings);
        combined.addAll(result.warnings());
        return new ValidationResult(result.errors(), combined);
    }
}

File: reference-implementations/dotnet/TomlSchema/DiscoveredSchema.cs
namespace TomlSchema;

using Tomlyn.Model;

/// <summary>
/// The outcome of discovering a schema from a TOML document via its reserved
/// <c>[toml-schema].location</c>, as described in SPEC.md's
/// "TOML Reference of a TOML Schema" section.
/// </summary>
/// <param name="Schema">The discovered and loaded schema.</param>
/// <param name="Document">The parsed TOML document that referenced the schema.</param>
/// <param name="Warnings">
/// Non-fatal version-compatibility warnings produced while comparing the document's
/// expected TOML Schema version against the resolved schema's declared version; empty
/// when the document omits <c>[toml-schema].version</c> or the versions match exactly.
/// </param>
public sealed record DiscoveredSchema(
    TomlSchema Schema,
    TomlTable Document,
    IReadOnlyList<ValidationDiagnostic> Warnings)
{
    /// <summary>
    /// Validates the discovered document against the discovered schema, including any
    /// version-compatibility warnings produced during discovery.
    /// </summary>
    public ValidationResult Validate()
    {
        var result = Schema.Validate(Document);
        if (Warnings.Count == 0)
            return result;

        var combined = new List<ValidationDiagnostic>(Warnings);
        combined.AddRange(result.Warnings);
        return new ValidationResult(result.Errors, combined);
    }
}
